### PR TITLE
[RISCV][GISel] Add ZFA FP legalization and full tests for 9 insn

### PR DESCRIPTION
**Pull request by lquinn2015**

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 1]
```diff
@@ -582,10 +582,13 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
...
 
   // FIXME: We can do custom inline expansion like SelectionDAG.
-  // FIXME: Legal with Zfa.
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
-                               G_INTRINSIC_ROUNDEVEN})
+                               G_INTRINSIC_ROUNDEVEN, G_FMAXIMUM, G_FMINIMUM})
+        
+      .legalFor(ST.hasStdExtZfa() /*ST.hasStdExtF()*/, {s32})
+      .legalFor(ST.hasStdExtZfa() & ST.hasStdExtD()  , {s64})
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:590]: please use &&

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 2]
```diff
@@ -582,10 +582,13 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       .libcallFor(ST.is64Bit(), {{s32, s128}, {s64, s128}});
 
   // FIXME: We can do custom inline expansion like SelectionDAG.
-  // FIXME: Legal with Zfa.
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
-                               G_INTRINSIC_ROUNDEVEN})
+                               G_INTRINSIC_ROUNDEVEN, G_FMAXIMUM, G_FMINIMUM})
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:587]: Do we have tests for libcalls for G_FMAXIMUM, G_FMINIMUM without Zfa?

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll [hunk 3]
```diff
@@ -0,0 +1,78 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:1]: This tests more than the legalizer. Put it in the llvm/test/CodeGen/RISCV/GlobalISel/ directory and name it half-zfa.ll like the SelectionDAG test.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll
Comment (mshockwave) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:1]: please test legalizer with MIR tests, just like other tests in the same folder

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+; RUN: llc -mtriple=riscv64 -mattr=+zfa,+zfh -global-isel < %s \
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f16
+
+
+define half @fceil(half %a) {
+; CHECK-ZFA-f16-LABEL: fceil:
+; CHECK-ZFA-f16:       # %bb.0:
+; CHECK-ZFA-f16-NEXT:    fround.h fa0, fa0, rup
+; CHECK-ZFA-f16-NEXT:    ret
+    %b = call half @llvm.ceil.f16(half %a)
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:14]: I believe we usually use 2 space indentation for IR

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:1]: I don't think there's much value in that at this point now that we have pretty good support for scalar. End to end IR tests cover legalizer, regbank, and isel much more efficiently.

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 2 above)
Comment (lquinn2015) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:587]: No we do not and you also cannot legalize these via libcall either it seems that is a defect :/  

was trying with this snippet, I suppose this change needs more work. I will shift this to a draft

llc mtriple=riscv32 -mattr+f -global-isel %s  
```
define float @fmaximum(float %a, float %b) {
   %c = call float @llvm.maximum.f32(float %a, float %b)
    ret float %c
}
```

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:587]: Just add a separate `getActionDefinitionsBuilder({G_FMAXIMUM, G_FMINIMUM})` that doesn't contain the libcallFor.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll (hunk 4 above)
Comment (lquinn2015) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll]: yes this is true

File: llvm/test/CodeGen/RISCV/GlobalISel/float-zfa.ll [hunk 5]
```diff
@@ -0,0 +1,161 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+
+; RUN: llc -mtriple=riscv32 -mattr=+zfa -global-isel < %s \
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f32
+; RUN: llc -mtriple=riscv64 -mattr=+zfa -global-isel < %s \
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f32
+; RUN: llc -mtriple=riscv32 -mattr=+f -global-isel < %s \
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/float-zfa.ll:7]: We should already have coverage without Zfa in float-intrinsics.ll

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll
Comment (lquinn2015) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:1]: Yeah recent push moved this to  IR .ll tests legalization is just a problem in the libcall case for the min/max imum cases gotta figure that out

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 1 above)
Comment (lquinn2015) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp]: Sure no problem fixed in repush. I am going to rebase the changes to one commit when I figure out the minimum and maximum stuff

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll
Comment (topperc) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll:1]: We don't need to solve the min/max libcall in this patch. We can just enabled min/max with Zfa and leave the non-Zfa case for later.

File: llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll (hunk 3 above)
Comment (lquinn2015) [reply llvm/test/CodeGen/RISCV/GlobalISel/legalizer/legalize-zfa-f16.ll]: Moved this I temporarly disabled the Fmaximum and Fminimum tests for consistency I think we can still test the lib call cases for s32 and s64 cases but I need to figure out the libcall pattern for it

**Comment by lquinn2015** [comment]
Still failing the MIR check validaton for G_MINIMUM and G_MAXIMUM i don't know why perhaps my test isn't covering something up

File: llvm/test/CodeGen/RISCV/GlobalISel/double-zfa.ll [hunk 6]
```diff
@@ -0,0 +1,79 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+
+; RUN: llc -mtriple=riscv32 -mattr=+zfa,d -global-isel < %s \
+; RUN: | FileCheck %s --check-prefixes=CHECK-ZFA-f64
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/double-zfa.ll:4]: You can drop --check-prefixes and let it use the default of `CHECK`

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 7]
```diff
@@ -582,12 +582,19 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       .libcallFor(ST.is64Bit(), {{s32, s128}, {s64, s128}});
 
   // FIXME: We can do custom inline expansion like SelectionDAG.
-  // FIXME: Legal with Zfa.
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
                                G_INTRINSIC_ROUNDEVEN})
+      .legalFor(ST.hasStdExtZfa() /*ST.hasStdExtF()*/, {s32})
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:588]: Is it intentional for ST.hasStdExtF to be commented out? I am a little confused on its meaning as a comment.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:588]: I'd probably drop it, or change it to `/*hasStdExtF() is implied by hasStdExtZfa*/`

Comment (lquinn2015) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:588]: I'm going to drop it, comments arent checked by the compiler its probably just more confusing to include it

File: llvm/test/CodeGen/RISCV/GlobalISel/double-zfa.ll (hunk 6 above)
Comment (lquinn2015) [reply llvm/test/CodeGen/RISCV/GlobalISel/double-zfa.ll]: Dropped additional check prefix

File: llvm/test/CodeGen/RISCV/GlobalISel/float-zfa.ll (hunk 5 above)
Comment (lquinn2015) [reply llvm/test/CodeGen/RISCV/GlobalISel/float-zfa.ll]: Removed the intrinsic check here

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp (hunk 2 above)
Comment (lquinn2015) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp]: Yeah this is the best choice for now we can hand the libcall in a different patch set

**Comment by lquinn2015** [comment]
@topperc  - I updated the legalizer check i did not include the ZFA extension in the Sweep. Rather I added the alias opcode. Let me know if that makes sense. If we add the extension it passes the extra rules as "Oks"

File: llvm/test/CodeGen/RISCV/GlobalISel/half-zfa.ll [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+define half @froundeven(half %a) {
+; CHECK-LABEL: froundeven:
+; CHECK:       # %bb.0:
+; CHECK-NEXT:    fround.h fa0, fa0, rne
+; CHECK-NEXT:    ret
+  %b = call half @llvm.roundeven.f16(half %a)
+  ret half %b
+}
+define half @fmaximum(half %a, half %b) {
```
Comment (topperc) [inline llvm/test/CodeGen/RISCV/GlobalISel/half-zfa.ll:62]: Missing blank line here

Comment (lquinn2015) [reply llvm/test/CodeGen/RISCV/GlobalISel/half-zfa.ll:63]: I will do better with the formatting next time! but yes this is now fixed

File: llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp [hunk 9]
```diff
@@ -582,12 +582,19 @@ RISCVLegalizerInfo::RISCVLegalizerInfo(const RISCVSubtarget &ST)
       .libcallFor(ST.is64Bit(), {{s32, s128}, {s64, s128}});
 
   // FIXME: We can do custom inline expansion like SelectionDAG.
-  // FIXME: Legal with Zfa.
   getActionDefinitionsBuilder({G_FCEIL, G_FFLOOR, G_FRINT, G_FNEARBYINT,
                                G_INTRINSIC_TRUNC, G_INTRINSIC_ROUND,
```
Comment (topperc) [inline llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:586]: There are no tests for G_INTRINSIC_TRUNC

Comment (lquinn2015) [reply llvm/lib/Target/RISCV/GISel/RISCVLegalizerInfo.cpp:586]: This should be handle now. 

I have 9 instrinsics  9 tests by 3 types for each possible size

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**